### PR TITLE
feat: adjust max io requests when read block data to avoid oom

### LINKED ISSUE #8721: feat: adjust max_storage_io_requests with block size and memory size
**Summary**

`max_storage_io_requests` default is 1000. If we run 
```
insert into log select * from log;
```
it's oom because we fetch `1000 * block size` > memory size, adjust `max_storage_io_requests` as:
`memory size / block size`

### PR DESCRIPTION
**Pull request by BohuTANG**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Change the read block data max_io_requests to:
```
// Assume 160MB one block file.
let block_file_size = 160 * 1024 * 1024_usize;
let table_column_len = self.table_info.schema().fields().len();
let per_column_bytes = block_file_size / table_column_len;
let column_memory_usage = per_column_bytes * projection.len();
let max_io_requests = max_memory_usage / column_memory_usage
```

* `max_storage_max_io_requests` only controls the non-block data read IO requests, such as segment/index files read.
* `set max_memory_usage= xx` to control the read block IO requests.

`insert into log select * from log;` 25GB memory usage avg.
image

`select * from log ignore_result`:
image

Closes #8721

### CODE REVIEW

**Review by dantengsky — APPROVED** [review]
LGTM

**Comment by TCeason** [comment]
Maybe this pr can also fix this oom kill? https://github.com/datafuselabs/databend/issues/8615

**Comment by Xuanwo** [comment]
stateful tests failed for timeout.

File: src/query/settings/src/lib.rs [hunk 1]
```diff
@@ -105,13 +108,26 @@ impl Settings {
...
-                    num_cpus::get() as u64
-                } else {
-                    conf.query.num_cpus
-                };
-                ret.set_max_threads(cpus)?;
+            {
+                if ret.get_max_threads()? == 0 {
+                    let cpus = if conf.query.num_cpus == 0 {
+                        sysinfo::System::new_with_specifics(
+                            RefreshKind::new().with_cpu(CpuRefreshKind::everything()),
```
Comment (sundy-li) [inline src/query/settings/src/lib.rs:115]: num_cpus has a physical cpu num method.

File: src/query/storages/fuse/fuse/src/operations/read_data.rs [hunk 2]
```diff
@@ -114,13 +124,36 @@ impl FuseTable {
...
+            }
+            ReadDataKind::BlockDataAdjustIORequests => {
+                // Assume 160MB one block file.
+                let block_file_size = 160 * 1024 * 1024_usize;
+
+                let table_column_len = self.table_info.schema().fields().len();
+                let per_column_bytes = block_file_size / table_column_len;
+                let column_memory_usage = per_column_bytes * projection.len();
+                let max_memory_usage = ctx.get_settings().get_max_memory_usage()? as usize;
+                max_memory_usage / column_memory_usage
```
Comment (sundy-li) [inline src/query/storages/fuse/fuse/src/operations/read_data.rs:145]: should check whether it's in a proper range, eg:  `[1, 1000]`

**Review by zhang2014 — APPROVED** [review]

File: src/query/settings/src/lib.rs (hunk 1 above)
Comment (BohuTANG) [reply src/query/settings/src/lib.rs:115]: I am not sure the physical CPU is right for some containers.

Comment (BohuTANG) [reply src/query/settings/src/lib.rs:115]: We used `num_cpus::get()` as available CPUs, not physical cores.

**Pull request merged**